hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/auth/SignerFactory.java

```diff
@@ -0,0 +1,150 @@
+
+import software.amazon.awssdk.auth.signer.Aws4Signer;
+import software.amazon.awssdk.auth.signer.Aws4UnsignedPayloadSigner;
+import software.amazon.awssdk.auth.signer.AwsS3V4Signer;
+import software.amazon.awssdk.core.exception.SdkException;
+import software.amazon.awssdk.core.signer.NoOpSigner;
+import software.amazon.awssdk.core.signer.Signer;
+
+import java.io.IOException;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.hadoop.fs.s3a.S3AUtils.getConstructor;
+import static org.apache.hadoop.fs.s3a.S3AUtils.getFactoryMethod;
+import static org.apache.hadoop.fs.s3a.S3AUtils.translateException;
+
+/**
+ * Signer factory used to register and create signers.
+ */
+public final class SignerFactory {
+
+  public static final String VERSION_FOUR_SIGNER = "AWS4SignerType";
+  public static final String VERSION_FOUR_UNSIGNED_PAYLOAD_SIGNER = "AWS4UnsignedPayloadSignerType";
+  public static final String NO_OP_SIGNER = "NoOpSignerType";
+  private static final String S3_V4_SIGNER = "AWSS3V4SignerType";
+  static final String INSTANTIATION_EXCEPTION
+      = "instantiation exception";
+
+  private static final Map<String, Class<? extends Signer>> SIGNERS
+      = new HashMap<>();
+
+  static {
+    // Register the standard signer types.
+    SIGNERS.put(VERSION_FOUR_SIGNER, Aws4Signer.class);
+    SIGNERS.put(VERSION_FOUR_UNSIGNED_PAYLOAD_SIGNER, Aws4UnsignedPayloadSigner.class);
+    SIGNERS.put(NO_OP_SIGNER, NoOpSigner.class);
+    SIGNERS.put(S3_V4_SIGNER, AwsS3V4Signer.class);
+  }
+
+
+  private SignerFactory() {
+  }
+
+  /**
+   * Register an implementation class for the given signer type.
+   *
+   * @param signerType  The name of the signer type to register.
+   * @param signerClass The class implementing the given signature protocol.
+   */
+  public static void registerSigner(
+      final String signerType,
+      final Class<? extends Signer> signerClass) {
+
+    if (signerType == null) {
+      throw new IllegalArgumentException("signerType cannot be null");
+    }
+    if (signerClass == null) {
+      throw new IllegalArgumentException("signerClass cannot be null");
+    }
+
+    SIGNERS.put(signerType, signerClass);
+  }
+
+  /**
+   * Check if the signer has already been registered.
+   * @param signerType
+   */
+  public static void getSigner(String signerType) {
+    Class<? extends Signer> signerClass = SIGNERS.get(signerType);
+    if (signerClass == null) {
+      throw new IllegalArgumentException("unknown signer type: " + signerType);
+    }
+  }
+
+
+  /**
+   * Create an instance of the given signer.
+   *
+   * @param signerType The signer type.
+   *
+   * @return The new signer instance.
+   */
+  public static Signer createSigner(String signerType) throws IOException {
+    Class<? extends Signer> signerClass = SIGNERS.get(signerType);
+    Signer signer = null;
```

Comment:
From here on, this looks like it could be a generic method in S3AUtils. It would also avoid having to make the reflection helpers public.

Only not sure what the TODO on line 125 means.